File: puzzles/python3/the_gift.py
budgets = []
total_budget = 0

nb_participants = int(input())
gift_price = int(input())
for i in range(nb_participants):
    budget = int(input())
    budgets.append(budget)
    total_budget += budget

if (total_budget < gift_price):
    print("IMPOSSIBLE")
else:
    budgets.sort()
    for i in range(nb_participants):
        contribution = budgets[i]
        max_contribution = gift_price // (nb_participants - i)
        if contribution >= max_contribution:
            contribution = max_contribution
        print(contribution)
        gift_price -= contribution

File: puzzles/kotlin/src/the-gift.kt
import java.util.Scanner

fun main(args : Array<String>) {
  val input = Scanner(System.`in`)
  val n = input.nextInt()
  var price = input.nextInt()
  val budgets = ArrayList<Int>(n)
  var totalBudget = 0

  for (i in 0 until n) {
    val budget = input.nextInt()
    budgets.add(budget)
    totalBudget += budget
  }

  if (totalBudget < price) {
    println("IMPOSSIBLE")
  } else {
    budgets.sort()
    for (i in 0 until n) {
      var contribution = budgets[i]
      val maxContribution = price / (n - i)
      if (contribution >= maxContribution) {
        contribution = maxContribution
      }
      println(contribution)
      price -= contribution
    }
  }
}
